python/paddle/v2/fluid/backward.py

```diff
@@ -340,3 +380,147 @@ def append_backward(loss, parameter_list=None, no_grad_set=None):
         else:
             params_and_grads.append((param_var, None))
     return params_and_grads
+
+
+def _as_list(x):
+    if x is None:
+        return []
+    return list(x) if isinstance(x, collections.Sequence) else [x]
+
+
+def _find_op_path_(block, outputs, inputs, no_grad_set):
+    """
+    no_grad_set will also be changed
+    """
+    input_names = set([inp.name for inp in inputs])
+    output_names = set([out.name for out in outputs])
+
+    relevant_op_flags = [True] * len(block.ops)
+
+    # All the inputs of the block are used if inputs is empty,
+    if inputs:
+        for i, op in enumerate(block.ops):
+            if _some_in_set_(op.desc.input_arg_names(), input_names):
+                for name in op.desc.output_arg_names():
+                    if name not in no_grad_set:
+                        input_names.add(name)
+            else:
+                relevant_op_flags[i] = False
+
+    for i, op in reversed(list(enumerate(block.ops))):
+        if _some_in_set_(op.desc.output_arg_names(), output_names):
+            for name in op.desc.input_arg_names():
+                if name not in no_grad_set:
+                    output_names.add(name)
+        else:
+            relevant_op_flags[i] = False
+
+    op_path = [
+        block.ops[i] for i in range(len(block.ops)) if relevant_op_flags[i]
+    ]
+
+    if inputs:
+        for op in op_path:
+            for name in op.desc.input_arg_names():
+                if name not in input_names:
+                    no_grad_set.add(name)
+
+    return op_path
+
+
+def calc_gradient(targets, inputs, target_gradients=None, no_grad_set=None):
```

Comment:
I have a few questions about the `calc_gradient`:
1. It seems that this API appends partial backward to calculate gradients of `targets` to `inputs`. Can this API be used by users directly? And when will users use it?
2. Here are some special handlings(like `_rename_grad_`) to make sure this API can be invoked for multiple times. However, in each invoking the `program` itself will be changed(some backward ops are appended). So when we invoke it second this, the backward part generated in the first time is still in it and are regarded as forwarding part. Is this reasonable?

Reply:
1. Yes. Some algorithm have its own way of calculating gradients. One example is [Learning to learn by gradient descent by gradient descent](https://arxiv.org/abs/1606.04474).
2. Yes. This happens in cases where gradients from different parts needs to be combined in some way to calculate the final result.

Reply:
Got it!